### ISSUE TITLE
Graph/plot visualizers

### ISSUE THREAD

**Issue by int19h** [Microsoft/PTVS#748]
In any context where expression evaluation is available (i.e. REPL and debugging), it should be possible to use matplotlib facilities to render a plot using the current data from that context, in such a way that it can be preserved, monitored for changes, and compared to other plots side by side.

- One VS tool window per plot, non-modal, dynamically resizable and dockable.
- All usual matplotlib facilities should be available with no or minimal syntactic differences from regular Python plotting.
- Simple way to create a basic plot from a given variable or expression (e.g. context menu command in Locals/Watch/Autos/...)
- Vectorized output with full-featured zooming and spanning.
- Live updates as underlying data changes (stepping in debugger, changing state in REPL etc); with an option to disable

![image](https://cloud.githubusercontent.com/assets/1175142/9265155/1ad8f19a-41e4-11e5-9941-89de80affe25.png)

**Comment by willyd** [comment]
+1 for this.

**Comment by denfromufa** [comment]
@int19h did you have a chance to look at Rodeo REPL? there is a lot of overlap between spyder, ptvs, jupyter, and rodeo. not sure if some components can be reused between projects?

**Comment by zooba** [comment]
@willyd Can you expand on that at all? Say we could only do half the things listed in the original post - which ones would be most valuable to you?

**Comment by denfromufa** [comment]
Sorry for getting in again, but thinking about this, the thing that could distinguish PTVS on MPL plots is interactive editor for options of plot just like in Excel charts. So fine finish or explore the plot without re-evaluating the plot with modified options from REPL.
In fact most of programming consultants I know would prefer Excel charts because of interactivity.

**Comment by willyd** [comment]
@zooba Sure. What I am really missing is the same kind of interactivity you get with MATLAB. For example, with MATLAB when you debug a script that creates and updates figures and graphics you can still create new figures in the REPL to inspect local or global variables. To my knowledge this is impossible with PTVS. I know we can do inline graphs in the REPL while debugging but I am looking for something more interactive. This discussion came up before see https://pytools.codeplex.com/discussions/464415. 

I am also a huge fan of the ImageWatch plugin for C++, something along those lines would also be helpful. But that does not allow plotting and creating other types of graphics. 

Finally, I agree that a fully vectorized implementation would be ideal but I'd rather have an interface compatible with matplotlib so that I can get going fast.

Again, thank you for this great product!

**Comment by denfromufa** [comment]
@willyd @zooba 

it is possible to plot multiple non-blocking blocks of matplotlib in jupyter REPL with with PTVS, the important part is enabling QT GUI backend:

```
 import matplotlib.pyplot as plt
%matplotlib qt
plt.plot(range(5))
plt.figure()
plt.plot(range(3,0,-1))
```

**Comment by willyd** [comment]
@denfromufa Yes I knew that. However, to my knowledge it is not possible to create multiple detached non-blocking plots in the Debug Interactive REPL if you also create some plots in your script. I believe this happens because we are calling the Qt main loop from two different threads.

**Comment by denfromufa** [comment]
The following script runs fine without debugging mode, but the plots "hang" with PTVS debugging mode:

```
import matplotlib.pyplot as plt
plt.ion()
plt.plot(range(5))
plt.figure()
plt.plot(range(3,0,-1))
raw_input()
```

**Comment by int19h** [comment]
Our intent with the plots is to use our own matplotlib rendering backend to render the plot into XAML, such that it can then be displayed by VS itself without Qt (or any other UI with a main loop inside the debuggee) being used at all. So, absolutely, having these plots in _non-modal_ windows is part of the plan, as is using matplotlib to drive this all - we're not going to reinvent the wheel.

**Comment by willyd** [comment]
@int19h Sounds exactly like what I want.

**Comment by denfromufa** [comment]
Wpf?

**Comment by int19h** [comment]
Yep. Since most of UI within VS is already WPF, it's easier. Though there's also the possibility of having matplotlib output SVG and then render that - but then we're limited by the quality of SVG rendering in IE that is installed on that machine...

**Comment by zooba** [comment]
FWIW, last time I spoke with one of the matplotlib maintainers, he said they'd be happy to accept a XAML backend into the core library.

**Comment by denfromufa** [comment]
There is also mplexporter used by d3, bokeh, and plotly.

On rendering side there is oxyplot for wpf.
